Add ImagePNG.string_of_png, which allows the user to get the encoded …

berke opened the pull request:
…representation of a PNG image as a string.

Commit:
Add ImagePNG.string_of_png, which allows the user to get the encoded representation of a PNG image as a string.

rlepigre commented on imageLib.mli:
```diff
@@ -47,6 +47,7 @@ module PNG :
     module ReadPNG : ReadImage
 
     val write_png : string -> image -> unit
+    val string_of_png : image -> string
```
I'm not sure if I like the name of this function, because it is not specific to the PNG images (it can take ANY image to a PNG-encoded string). I suggest `to_png_string`, but I'll let you decide.

cfcs replied:
a docstring `(** [string_of_png img] is [img] encoded to PNG format as a string. *)` or something like that would be nice. Maybe also mentioning the OUT_CHANNEL stuff you added? (which I think is nice!)

cfcs replied:
@rlepigre Writing *any* `image` to a PNG seems to me like a useful thing (so you can convert between formats)? Is there a reason this is a bad idea?

rlepigre replied:
@cfcs of course no, the problem is that `string_of_png` suggests that you will convert a `png` into a `string`. But what this function does is convert an `image` (PNG or not) into the `string` representation of a PNG.

rlepigre commented on jbuild:
```diff
@@ -0,0 +1,6 @@
+(jbuild_version 1)
```
This file should not be part be added.

rlepigre commented on imagePNG.ml:
```diff
@@ -989,6 +989,21 @@ module ReadPNG : ReadImage = struct
      | _ -> assert false
 end
 
+module type OUT_CHANNEL = sig
```
Could you change the indentation to the following (for consistency with the beginning of the file).
```ocaml
module type OUT_CHANNEL = sig
  type out_channel
  val output_string : out_channel -> string -> unit
  val output_char : out_channel -> char -> unit
end

module Buffer_channel = struct
  type out_channel = Buffer.t
  let output_string = Buffer.add_string
  let output_char = Buffer.add_char
end
```

cfcs replied:
I think this module / interface should be defined somewhere else, since it'll likely be useful for the other image formats too.

rlepigre replied:
@cfcs you are right, but I was thinking of merging first, and then take care of moving things myself. I want to go over the interface of the library to add the missing documentation stuff.

rlepigre reviewed (commented):
Thanks for the contribution! I only have a few suggestions before I merge, the most important one being to remove the `jbuild` file. Also, could you add your name to the `CONTRIBUTIONS` file? There are more and more contributors to `imagelib` and I'd like to keep track of who did what.

cfcs commented on imagePNG.ml:
```diff
@@ -1246,6 +1260,21 @@ let write_png fn img =
 
-  close_out och
+let string_of_png img =
+  let approx_size = img.width * img.height in
```
`Buffer` is limited to `Sys.max_string_size` which is pretty small on some platforms (notably 16 MB on x86).
You need to catch the exception when the internal buffer would exceed this, or document the potential exception in the `mli`.

Commit:
Move OUT_CHANNEL and Buffer_channel to an ImageChannels module, and replace string_of_png with the more logical bytes_of_png.

berke commented:
Hi,

I renamed string_of_png to bytes_of_png, put the channel stuff in a separate imageChannels module and added an entry in CONTRIBUTIONS.  I don't know how to easily overlay the jbuild file which I need, so I'm leaving it for now as you will be reorganizing things anyway.

BTW thanks for this nice library and especially the PNG implementation.

If you do reorganize at some point maybe you will add a WriteImage module signature similar to the ReadImage.  I only added the bare minimum to get PNG writing capability (we're using it in a little HTTP server.)

rlepigre commented:
Cool, I'll merge now, but can you elaborate on why you need the `jbuild` file?

Thanks again for the contribution!

berke commented:
> Cool, I'll merge now, but can you elaborate on why you need the jbuild file?

Thanks.

As a temporary measure to ease development.  Normally we use imagelib via opam.  At first I was using an opam pin to compile it with my changes, but I found that to be slow and inconvenient.  So I placed imagelib as a git submodule under our work tree and I added a jbuild so that it could be compiled quickly with the rest.  When the work is done, I'll remove the submodule and use an opam dependency.

It's possible I wasn't using opam optimally.  Is there a better way?

cfcs commented:
I think unfortunately opam is just a bit slow :(
